### ISSUE TITLE
Version nuget task failed to match 5 digits for the revision number

### ISSUE THREAD

**Issue by yanlinsun** [rfennell/AzurePipelines#993]
### Azure DevOps Extensions

Manifest Versioning

### Platform

Azure DevOps Services

### Extension Version

2.29.16

### Describe the bug

```bash
The default regular expression failed to match the 5 digits for the revision number.
If the target build number is 2.0.0.12345, the version nuspec package task only extract 
2.0.0.1234
```

### Repo Steps

```bash
1. Use 'Version nuspec' task in build pipeline
2. Ensure the build number is "2.0.0.12345" when the version nuspec task is executing.
3. Check the log, you will find the extracted version is "2.0.0.1234".
```

### Expected Behavior

```bash
The version should be "2.0.0.12345".


By analyzing the regular expression, in the expression '0|[1-9][0-9]{0,3}|[1-5][0-9]{4}|6[0-4][0-9]{3}|65[0-4][0-9]{2}|655[0-2][0-9]|6553[0-5]', the second alternative '[1-9][0-9]{0,3}' has higher priority to the following alternatives.
So, when it finds a match it stops. 
An easy way to overcome this is to revert the expression as '6553[0-5]|655[0-2][0-9]|65[0-4][0-9]{2}|6[0-4][0-9]{3}|[1-5][0-9]{4}|[1-9][0-9]{0,3}|0'.
```

### Logging Information

```bash
##[debug]Evaluating condition for step: 'Version Nuspec'
##[debug]Evaluating: succeeded()
##[debug]Evaluating succeeded:
##[debug]=> True
##[debug]Result: True
Starting: Version Nuspec
==============================================================================
Task         : Version Nuspec package
Description  : Applies a version to a Nuspec manifest based on the build number. Based on Microsoft sample from https://msdn.microsoft.com/Library/vs/alm/Build/overview
Version      : 2.29.16
Author       : Black Marble
Help         : Version: 2.29.16. [More Information](https://github.com/rfennell/vNextBuild/wiki/Version-Assemblies-and-Packages-Tasks/)
==============================================================================
##[debug]VstsTaskSdk 0.10.0 commit 787a46ec0a2df5b4d12c2e801bd3f319975c054c
##[debug]INPUT_PATH: 'C:\agent1\_work\80\s'
##[debug]INPUT_VERSIONNUMBER: '3.2.3.21105-task-11882-api-nuget-auto-publish'
##[debug]INPUT_INJECTVERSION: 'False'
##[debug]INPUT_VERSIONREGEX: '(0|[1-9][0-9]{0,3}|[1-5][0-9]{4}|6[0-4][0-9]{3}|65[0-4][0-9]{2}|655[0-2][0-9]|6553[0-5])(\.(0|[1-9][0-9]{0,3}|[1-5][0-9]{4}|6[0-4][0-9]{3}|65[0-4][0-9]{2}|655[0-2][0-9]|6553[0-5])){3}'
##[debug]INPUT_OUTPUTVERSION: 'OutputedVersion'
##[debug]INPUT_FILENAME: 'Flintfox.nPrice.Api.nuspec'
##[debug]Source Directory: C:\agent1\_work\80\s
##[debug]Filename: Flintfox.nPrice.Api.nuspec
##[debug]Version Number/Build Number: 3.2.3.21105-task-11882-api-nuget-auto-publish
##[debug]Version Filter: (0|[1-9][0-9]{0,3}|[1-5][0-9]{4}|6[0-4][0-9]{3}|65[0-4][0-9]{2}|655[0-2][0-9]|6553[0-5])(\.(0|[1-9][0-9]{0,3}|[1-5][0-9]{4}|6[0-4][0-9]{3}|65[0-4][0-9]{2}|655[0-2][0-9]|6553[0-5])){3}
##[debug]Inject Version: False
##[debug]Output: Version Number Parameter Name: OutputedVersion
##[debug]Version: 3.2.3.2110
##[debug]Will apply 3.2.3.2110 to 1 files.
##[debug]Updated the file C:\agent1\_work\80\s\net 5\Flintfox.nPrice.Api\Flintfox.nPrice.Api.nuspec with the version 3.2.3.2110
##[debug]Set the output variable 'OutputedVersion' with the value 3.2.3.2110
##[debug]Processed: ##vso[task.setvariable variable=OutputedVersion;]3.2.3.2110
```

**Comment by rfennell** [comment]
This regex value is passed in a parameter so you can swap to anyone you prefer.

I will however  look to getting the default changed on this and the other tasks in this extension that use the same expression

**Comment by rfennell** [comment]
I have had a chance to look at the code, and the regex match used is the zeroth match entry in the array,

```
    $VersionData = [regex]::matches($VersionNumber,$VersionRegex)
    switch($VersionData.Count)
    {
    0        
        { 
            Write-Error "Could not find version number data in $VersionNumber."
            exit 1
        }
    1 {}
    default 
        { 
            Write-Warning "Found more than instance of version data in $VersionNumber." 
            Write-Warning "Will assume first instance is version."
        }
    }
    $NewVersion = $VersionData[0]
```

Using a [regex tester](https://regex101.com/), if I swap to using your suggested `6553[0-5]|655[0-2][0-9]|65[0-4][0-9]{2}|6[0-4][0-9]{3}|[1-5][0-9]{4}|[1-9][0-9]{0,3}|0` this code does not find the required match.

However, by swapping the quantifiers from `{0, 3}` to `{0, 4}` it seems to do the match for me i.e.

```
# Version Nuspec package
# Description - Applies a version to a Nuspec manifest based on the build number. Based on Microsoft sample from https://msdn.microsoft.com/Library/vs/alm/Build/overview
- task: VersionNuspec@2
  inputs: 
     # Required arguments
     Path: $(Build.SourcesDirectory)
     Filename: *.nuspec
     VersionNumber: $(Build.BuildNumber)
     InjectVersion: False
     VersionRegex: (0|[1-9][0-9]{0,4}|[1-5][0-9]{4}|6[0-4][0-9]{3}|65[0-4][0-9]{2}|655[0-2][0-9]|6553[0-5])(\.(0|[1-9][0-9]{0,4}|[1-5][0-9]{4}|6[0-4][0-9]{3}|65[0-4][0-9]{2}|655[0-2][0-9]|6553[0-5])){3}
     OutputVersion: OutputedVersion
``

As the `VersionRegex` is a parameter, and this is the first reported issue in all the years the task has been available, I do not intend to alter the default expression. You can of course alter the parameter for your pipeline

**Comment by yanlinsun** [comment]
Yes, I altered my pipeline to use a regex as simple as '\d+(\.\d+){3}'. This is to label my NuGet package version and base on my testing it has no issue if any of the portions exceed 65535 (the DLL's version inside the NuGet package does not exceed 65535). 

BTW, the link https://msdn.microsoft.com/Library/vs/alm/Build/overview is no longer available.

I did a search on regular expression number matching and I found this article https://www.regular-expressions.info/numericranges.html.
The first section points out the same issue.

> Putting this all together using alternation we get: [0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5]. This matches the numbers we want, with one caveat: regular expression searches usually allow partial matches, so our regex would match 123 in 12345. There are two solutions to this.

The article suggests two solutions: 1. to use word boundaries \b, 2. to use anchors ^ and $. 
I hope this could help with your solution.
